src-opam/dockerfile_opam.ml

```diff
@@ -109,40 +109,90 @@ let header ?win10_revision ?arch ?maintainer ?img ?tag d =
   @@ maintainer
   @@ shell
 
+type opam_hashes = {
+  opam_2_0_hash : string;
+  opam_2_1_hash : string;
+  opam_master_hash : string;
+}
+
+type opam_branch = {
+  branch : string;
+  hash : string;
+  enable_0install_solver : bool;
+  public_names : string list;
+}
+
+let create_opam_branches opam_hashes =
+  let {
+    opam_2_0_hash;
+    opam_2_1_hash;
+    opam_master_hash;
+  } = opam_hashes in
+  [
+    {
+      branch = "2.0";
+      hash = opam_2_0_hash;
+      enable_0install_solver = false;
+      public_names = ["2.0"];
+    };
+    {
+      branch = "2.1";
+      hash = opam_2_1_hash;
+      enable_0install_solver = true;
+      public_names = ["2.1"];
+    };
+    {
+      branch = "master";
+      hash = opam_master_hash;
+      enable_0install_solver = true;
+      public_names = ["2.2"; "dev"]; (* TODO: Change when opam 2.2 is branched *)
+    };
+  ]
+
 (* Apk based Dockerfile *)
-let apk_opam2 ?(labels=[]) ?arch ~hash_opam_2_0 ~hash_opam_2_1 distro () =
+let apk_opam2 ?(labels=[]) ?arch ~opam_hashes distro () =
+  let opam_branches = create_opam_branches opam_hashes in
   let img, tag = D.base_distro_tag ?arch distro in
   header ?arch distro @@ label (("distro_style", "apk") :: labels)
   @@ Linux.Apk.install "build-base bzip2 git tar curl ca-certificates openssl"
   @@ Linux.Git.init ()
-  @@ install_opam_from_source ~add_default_link:false ~branch:"2.0" ~hash:hash_opam_2_0 ()
-  @@ install_opam_from_source ~add_default_link:false ~enable_0install_solver:true ~branch:"2.1" ~hash:hash_opam_2_1 ()
+  @@ List.fold_left (fun acc {branch; hash; enable_0install_solver; _} ->
+      acc @@ install_opam_from_source ~add_default_link:false ~enable_0install_solver ~branch ~hash ()
+    ) empty opam_branches
   @@ run "strip /usr/local/bin/opam*"
   @@ from ~tag img
   @@ Linux.Apk.add_repository ~tag:"edge" "https://dl-cdn.alpinelinux.org/alpine/edge/main"
   @@ Linux.Apk.add_repository ~tag:"edgecommunity" "https://dl-cdn.alpinelinux.org/alpine/edge/community"
   @@ Linux.Apk.add_repository ~tag:"testing" "https://dl-cdn.alpinelinux.org/alpine/edge/testing"
-  @@ copy ~from:"0" ~src:["/usr/local/bin/opam-2.0"] ~dst:"/usr/bin/opam-2.0" ()
-  @@ copy ~from:"0" ~src:["/usr/local/bin/opam-2.1"] ~dst:"/usr/bin/opam-2.1" ()
+  @@ List.fold_left (fun acc {branch; public_names; _} ->
+      List.fold_left (fun acc public_name ->
+        acc @@ copy ~from:"0" ~src:["/usr/local/bin/opam-"^branch] ~dst:("/usr/bin/opam-"^public_name) ()
+      ) acc public_names
```

Comment:
Wants some further tidying to eliminate the `assert false`, but this should shave 18M off the images (I don't think docker naturally de-dups?)
```suggestion
      match public_names with
      | [] -> assert false
      | main::aliases ->
          let main = "/usr/bin/opam-"^main in
          let acc = acc @@ copy ~from:"0" ~src:["/usr/local/bin/opam-"^branch] ~dst:main () in
          List.fold_left (fun acc public_name ->
            acc @@ run "ln %s /usr/bin/opam-%s" main public_name
          ) acc aliases
```

Reply:
Done